Integration test encryption homedir

PVince81 opened the pull request:
Extension of https://github.com/owncloud/core/pull/26586 but where the home directory name is different from the user id.

It currently uses a hack to md5 the user's home.

Goal is to reproduce the issue from https://github.com/owncloud/core/pull/26820 automatically.

Let's see how it goes...

Commit:
Add integration test stage with encryption enabled

Commit:
Add no_encryption filter for incompatible int tests

Added filter to disable some tests when encryption is enabled.

Commit:
Make sure we could create the work folder in int tests

PVince81 commented:
Ouch, many fails even before we even started enabling encryption:
```
    checksums.Moving file with checksum should return the checksum in the download header
    sharing.Share a file by multiple channels
    sharing.Delete all group shares
    sharing.Allow modification of reshare
    sharing.Do not allow reshare to exceed permissions
    sharing.Correct webdav share-permissions for received folder with all permissions
    sharing.Correct webdav share-permissions for received folder with all permissions but edit
    sharing.Correct webdav share-permissions for received folder with all permissions but create
    sharing.Correct webdav share-permissions for received folder with all permissions but delete
    sharing.Correct webdav share-permissions for received folder with all permissions but share
    sharing.Merging shares for recipient when shared from outside with user then group and recipient renames in between
    sharing.orphaned shares
    webdav-related.Moving a file to overwrite a file in a folder with no permissions
    webdav-related.Copying a file
    webdav-related.Copying a file to a folder with no permissions
    webdav-related.Copying a file to overwrite a file into a folder with no permissions
    webdav-related.Retrieving folder quota when quota is set
    webdav-related.Retrieving folder quota of shared folder with quota when no quota is set for recipient
    webdav-related.Uploading a file as recipient using webdav having quota
    webdav-related.Retrieving folder quota when quota is set and a file was uploaded
    webdav-related.Retrieving folder quota when quota is set and a file was recieved
    webdav-related.A file that is shared to a user has a share-types property
    webdav-related.A file that is shared to a group has a share-types property
    webdav-related.A file that is shared by link has a share-types property
    webdav-related.A file that is shared by user,group and link has a share-types property
    federated.Overwrite a federated shared file as recipient
    federated.Overwrite a federated shared folder as recipient
```

It could however be a side-effect.
I'll rerun the tests this time without the encryption-related changes and only with the md5'd home.

PVince81 commented:
I made a separate PR where only the home is MD5'd: https://github.com/owncloud/core/pull/26846
This will at least make it possible to work on other fixes until we decide how to make this testing fully automatic.

PVince81 commented:
- [ ] Requires https://github.com/owncloud/core/pull/26848 for having proper test cleanups

SergioBertolinSG commented:
I didn't see it finish in my local environment:

```
Scenario: Federate share a file with another server                                                    # /var/www/html/master_20-12-16/build/integration/federation_features/federated.feature:5
    Given Using server "REMOTE"                                                                          # FederationContext::usingServer()
./run.sh: line 40: 28695 Killed                  vendor/bin/behat --strict -f junit -f pretty $SCENARIO_TO_RUN
./run.sh: line 46: 28685 Terminated              php -S localhost:$PORT -t ../..
./run.sh: line 46: 28686 Terminated              php -S localhost:$PORT_FED -t ../..
files_external disabled
{"reqId":"wOVIiDwNAXh1kccM+Dsu","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:22+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToTag.txt","user":"--"}
{"reqId":"UqvcPZXoz78bKSyiVeIg","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:24+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToComment.txt","user":"--"}
{"reqId":"Dg2BJNO41njfpVz+w7ek","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:24+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToTag.txt","user":"--"}
{"reqId":"QJ7icWu7ynRA7DrPYqAm","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 503 OC\\\\User\\\\LoginException: User disabled\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\ServiceUnavailable\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(199): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Auth->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(150): Sabre\\\\DAV\\\\Auth\\\\Plugin->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#5 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(57): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/var\\\/www\\\/html\\\/master_20-12-16\\\/remote.php(164): require_once('\\\/var\\\/www\\\/html\\\/m...')\\n#8 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/master_20-12-16\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Auth.php\",\"Line\":153,\"User\":false}","level":4,"time":"2016-12-20T17:43:26+00:00","method":"GET","url":"\/remote.php\/webdav\/welcome.txt","user":"--"}
{"reqId":"aWxuucVr1aKMiFuGoBYb","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:26+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToComment.txt","user":"--"}
{"reqId":"O3p1TlNtXEmmTbihW4ld","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:27+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToTag.txt","user":"--"}
{"reqId":"aeV9rQNHhfXwR\/Tn\/zaB","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:28+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToComment.txt","user":"--"}
{"reqId":"OjrXH+Pgjm\/aXg0SnDm4","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:28+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToTag.txt","user":"--"}
{"reqId":"BrFOS61rGknOutgMafa0","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:30+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToComment.txt","user":"--"}
{"reqId":"nlzlhXdq3lMerdQn6pUX","remoteAddr":"127.0.0.1","app":"core","message":"Login failed: 'user0' (Remote IP: '127.0.0.1')","level":2,"time":"2016-12-20T17:43:30+00:00","method":"DELETE","url":"\/remote.php\/webdav\/myFileToTag.txt","user":"--"}
runsh: Exit code: 137
```

PVince81 commented:
Let me add that PR here. I hoped we could quickly get https://github.com/owncloud/core/pull/26848 merged and rebase.

PVince81 commented:
also I'd prefer if we could get https://github.com/owncloud/core/pull/26586 merged first which introduces the encryption tests in the first place.

PVince81 commented:
I've added https://github.com/owncloud/core/pull/26848 now here.

Ideally we shouldn't see more failures than in the encryption-less test here https://github.com/owncloud/core/pull/26846. But I expect we should see more failures, some being related to https://github.com/owncloud/core/pull/26820. If they don't, we'll need to add more tests to make sure to cover that PR's scope.

PVince81 commented:
Rebased and adjusted.

See how the test matrix for integration tests is exploding

PVince81 commented:
Argh, I need to remove the hack commit

PVince81 commented:
Fixes https://github.com/owncloud/QA/issues/324

PVince81 commented:
Now I hope that there is a failing test that matches the scenario from https://github.com/owncloud/core/pull/26820

PVince81 commented:
Aha, finally some results after running `OC_TEST_ENCRYPTION_ENABLED=1 OC_TEST_ALT_HOME=1 ./run.sh features/sharing-v1.feature` locally:

```
    /srv/www/htdocs/owncloud/build/integration/features/sharing-v1.feature:269
    /srv/www/htdocs/owncloud/build/integration/features/sharing-v1.feature:453
```

For the first one:
```
[Tue Jan 10 19:23:25 2017] 127.0.0.1:34530 [204]: /remote.php/webdav/textfile0%20(2).txt
    When User "user1" deletes file "/textfile0 (2).txt"                              # FeatureContext::userDeletesFile()
    And As an "user3"                                                                # FeatureContext::asAn()
[Tue Jan 10 19:23:26 2017] Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.
[Tue Jan 10 19:23:26 2017] Exception: {"Message":"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.","Exception":"OC\\Encryption\\Exceptions\\DecryptionFailedException","Code":0,"Trace":"#0 \/srv\/www\/htdocs\/owncloud\/lib\/private\/Files\/Stream\/Encryption.php(459): OCA\\Encryption\\Crypto\\Encryption->decrypt('hONVoITLn+\/yEvL...', '0end')\n#1 \/srv\/www\/htdocs\/owncloud\/lib\/private\/Files\/Stream\/Encryption.php(290): OC\\Files\\Stream\\Encryption->readCache()\n#2 [internal function]: OC\\Files\\Stream\\Encryption->stream_read(23)\n#3 \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/Wrapper.php(83): fread(Resource id #559, 8192)\n#4 \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/CallbackWrapper.php(91): Icewind\\Streams\\Wrapper->stream_read(8192)\n#5 [internal function]: Icewind\\Streams\\CallbackWrapper->stream_read(8192)\n#6 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/http\/lib\/Sapi.php(78): stream_copy_to_stream(Resource id #563, Resource id #573, '7')\n#7 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(498): Sabre\\HTTP\\Sapi::sendResponse(Object(Sabre\\HTTP\\Response))\n#8 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#9 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(57): Sabre\\DAV\\Server->exec()\n#10 \/srv\/www\/htdocs\/owncloud\/remote.php(164): require_once('\/srv\/www\/htdocs...')\n#11 {main}","File":"\/srv\/www\/htdocs\/owncloud\/apps\/encryption\/lib\/Crypto\/Encryption.php","Line":360,"User":"user3"}
[Tue Jan 10 19:23:26 2017] 127.0.0.1:34552 [500]: /remote.php/webdav/textfile0%20(2).txt
    And Downloading file "/textfile0 (2).txt" with range "bytes=1-7"                 # FeatureContext::downloadFileWithRange()
      Server error response [url] http://localhost:8080/remote.php/webdav/textfile0%20(2).txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)
    Then Downloaded content should be "wnCloud"                                      # FeatureContext::downloadedContentShouldBe()
```

and the second one:
```
[Tue Jan 10 19:24:27 2017] 127.0.0.1:35744 [201]: /remote.php/webdav/textfile0.txt
    And User "user1" moved file "/textfile0.txt" to "/common/textfile0.txt"                                     # FeatureContext::userMovedFile()
[Tue Jan 10 19:24:28 2017] Could not encrypt file for user2: Public Key missing for user: user2
[Tue Jan 10 19:24:28 2017] 127.0.0.1:35766 [201]: /remote.php/webdav/common/textfile0.txt
    And User "user1" moved file "/common/textfile0.txt" to "/common/sub/textfile0.txt"                          # FeatureContext::userMovedFile()
    And As an "user2"                                                                                           # FeatureContext::asAn()
[Tue Jan 10 19:24:28 2017] Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.
[Tue Jan 10 19:24:28 2017] Exception: {"Message":"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.","Exception":"OC\\Encryption\\Exceptions\\DecryptionFailedException","Code":0,"Trace":"#0 \/srv\/www\/htdocs\/owncloud\/lib\/private\/Files\/Stream\/Encryption.php(459): OCA\\Encryption\\Crypto\\Encryption->decrypt('lGceeUTJeo\/pGEs...', '0end')\n#1 \/srv\/www\/htdocs\/owncloud\/lib\/private\/Files\/Stream\/Encryption.php(290): OC\\Files\\Stream\\Encryption->readCache()\n#2 [internal function]: OC\\Files\\Stream\\Encryption->stream_read(15)\n#3 \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/Wrapper.php(83): fread(Resource id #561, 8192)\n#4 \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/CallbackWrapper.php(91): Icewind\\Streams\\Wrapper->stream_read(8192)\n#5 [internal function]: Icewind\\Streams\\CallbackWrapper->stream_read(8192)\n#6 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/http\/lib\/Sapi.php(78): stream_copy_to_stream(Resource id #565, Resource id #575, '9')\n#7 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(498): Sabre\\HTTP\\Sapi::sendResponse(Object(Sabre\\HTTP\\Response))\n#8 \/srv\/www\/htdocs\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#9 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(57): Sabre\\DAV\\Server->exec()\n#10 \/srv\/www\/htdocs\/owncloud\/remote.php(164): require_once('\/srv\/www\/htdocs...')\n#11 {main}","File":"\/srv\/www\/htdocs\/owncloud\/apps\/encryption\/lib\/Crypto\/Encryption.php","Line":360,"User":"user2"}
[Tue Jan 10 19:24:28 2017] 127.0.0.1:35788 [500]: /remote.php/webdav/common/sub/textfile0.txt
    When Downloading file "/common/sub/textfile0.txt" with range "bytes=9-17"                                   # FeatureContext::downloadFileWithRange()
      Server error response [url] http://localhost:8080/remote.php/webdav/common/sub/textfile0.txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)
    Then Downloaded content should be "test text"                                                               # FeatureContext::downloadedContentShouldBe()
    And Downloaded content when downloading file "/textfile0.txt" with range "bytes=9-17" should be "test text" # FeatureContext::downloadedContentWhenDownloadindShouldBe()
    And user "user2" should see following elements                                                              # FeatureContext::checkElementList()
      | /common/sub/textfile0.txt |
```

Now let's see if these pass with https://github.com/owncloud/core/pull/26824

PVince81 commented:
Yes, the tests from above pass if I cherry-pick the changes from https://github.com/owncloud/core/pull/26824 to this PR. This is what I wanted to cover.

PVince81 commented:
Fixed the Jenkinsfile to use the correct env variable.

@SergioBertolinSG please review this PR.

PVince81 commented on build/integration/Makefile:
```diff
@@ -22,5 +22,5 @@ clean-test-results:
 
 test: install-composer-deps
-	OC_TEST_ALT_HOME=$(OC_TEST_ALT_HOME) ./run.sh
+	OC_TEST_ALT_HOME=$(OC_TEST_ALT_HOME) OC_TEST_ENCRYPTION_ENABLED=$(ENCRYPTION_ENABLED) ./run.sh
```
argh, missed this one

PVince81 commented:
We need to merge https://github.com/owncloud/core/pull/26917 first, else we'd have CI failures on master.

- [ ] REQUIRES: https://github.com/owncloud/core/pull/26917

PVince81 commented:
Seems we have some additional failures (local run):
```
    /srv/www/htdocs/owncloud/build/integration/features/external-storage.feature:7
    /srv/www/htdocs/owncloud/build/integration/features/sharing-v1.feature:219
    /srv/www/htdocs/owncloud/build/integration/features/sharing-v1.feature:283
    /srv/www/htdocs/owncloud/build/integration/features/webdav-related.feature:197
    /srv/www/htdocs/owncloud/build/integration/features/webdav-related.feature:207
    /srv/www/htdocs/owncloud/build/integration/federation_features/federated.feature:123
    /srv/www/htdocs/owncloud/build/integration/federation_features/federated.feature:138
```

PVince81 commented:
- [ ] BUG: `/srv/www/htdocs/owncloud/build/integration/features/sharing-v1.feature:219` happens when encryption is enabled, works when encryption is off:
```
[Wed Jan 11 08:43:52 2017] 127.0.0.1:42350 [200]: /ocs/v1.php/apps/files_sharing/api/v1/shares
    When sending "GET" to "/apps/files_sharing/api/v1/shares"            # FeatureContext::sendingTo()
    Then the OCS status code should be "100"                             # FeatureContext::theOCSStatusCodeShouldBe()
    And the HTTP status code should be "200"                             # FeatureContext::theHTTPStatusCodeShouldBe()
    And File "textfile0.txt" should be included in the response          # FeatureContext::checkSharedFileInResponse()
      │ /textfile0 (2).txt
      Failed asserting that false matches expected true.
```

PVince81 commented:
- These work when run individually, so possibly a side-effect of the other failing tests:
    - /srv/www/htdocs/owncloud/build/integration/features/webdav-related.feature:197
    - /srv/www/htdocs/owncloud/build/integration/features/webdav-related.feature:207

- [ ] BUG: failure of federated feature has similar results with encryption:
    - /srv/www/htdocs/owncloud/build/integration/federation_features/federated.feature:123
    - /srv/www/htdocs/owncloud/build/integration/federation_features/federated.feature:138

```
[Wed Jan 11 08:54:34 2017] 127.0.0.1:49320 [200]: /ocs/v1.php/apps/files_sharing/api/v1/remote_shares/pending/1
    And User "user1" from server "REMOTE" accepts last pending share                                    # FederationContext::acceptLastPendingShare()
    And Using server "REMOTE"                                                                           # FederationContext::usingServer()
    And As an "user1"                                                                                   # FederationContext::asAn()
    And User "user1" modifies text of "/textfile0.txt" with text "BLABLABLA"                            # FederationContext::modifyTextOfFile()
      Warning: file_put_contents(../../data/user1/files/textfile0.txt): failed to open stream: No such file or directory in /srv/www/htdocs/owncloud/build/integration/features/bootstrap/BasicStructure.php line 303
    When User "user1" uploads file "../../data/user1/files/textfile0.txt" to "/textfile0 (2).txt"       # FederationContext::userUploadsAFileTo()
    And Downloading file "/textfile0 (2).txt" with range "bytes=0-8"                                    # FederationContext::downloadFileWithRange()
    Then Downloaded content should be "BLABLABLA"
```

PVince81 commented:
The reason for the "(2)" that gets added after the share is because that user already has a file called "textfile0.txt".

When encryption is disabled, creating the share doesn't automatically add the "(2)", but it would happen during the next setupFS for that user. So if I manually do a PROPFIND for that user, it will appear. That doesn't happen during the automated tests.

However when encryption is enabled, it seems that this logic is triggered earlier, so the user will get the "(2)" earlier. This explains the results above.

Now the part that bothers me is that this makes the results quite unstable. I think before querying the shares in these tests we should first do a PROPFIND for the recipient user and check if the file exists, and always expect the one with "(2)" to exist. Or, an alternative would be to create a new file name that has no conflicts at all to avoid the "(2)" issue altogether. (nevertheless we'll need additional specific tests to verify that the deduplication happens, if we don't have these already)

@SergioBertolinSG what do you think ?

PVince81 commented:
Rebased to integrate https://github.com/owncloud/core/pull/26920

Oh, surprisingly this seems to fix the sharing failure mentioned above.

I'll then redo a full run to check it.

PVince81 commented:
Hmmm no, I was wrong. On this PR here these tests will pass.

But as soon as I cherry-pick the fix from https://github.com/owncloud/core/pull/26917 they will fail because that fix contains additional `initMountPoints` calls.

PVince81 commented:
@SergioBertolinSG can we maybe rename the file before sharing to avoid the name clash in https://github.com/owncloud/core/pull/26844#issuecomment-271802108 ?

SergioBertolinSG commented:
There is an exception where a user existence is checked (lot of times):

```
[Thu Jan 12 08:18:44 2017] Exception: {"Exception":"OCA\\Encryption\\Exceptions\\PublicKeyMissingException","Message":"Public Key missing for user: subadmin","Code":0,"Trace":"#0 \/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/KeyManager.php(519): OCA\\Encryption\\KeyManager->getPublicKey('subadmin')\n#1 \/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/Users\/Setup.php(74): OCA\\Encryption\\KeyManager->userHasKeys('subadmin')\n#2 \/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/Hooks\/UserHooks.php(187): OCA\\Encryption\\Users\\Setup->setupUser('subadmin', '123456')\n#3 \/var\/www\/html\/encryption_vincent\/lib\/private\/legacy\/hook.php(105): OCA\\Encryption\\Hooks\\UserHooks->postCreateUser(Array)\n#4 \/var\/www\/html\/encryption_vincent\/lib\/private\/Server.php(253): OC_Hook::emit('OC_User', 'post_createUser', Array)\n#5 [internal function]: OC\\Server->OC\\{closure}(Object(OC\\User\\User), '123456')\n#6 \/var\/www\/html\/encryption_vincent\/lib\/private\/Hooks\/EmitterTrait.php(98): call_user_func_array(Object(Closure), Array)\n#7 \/var\/www\/html\/encryption_vincent\/lib\/private\/Hooks\/PublicEmitter.php(32): OC\\Hooks\\BasicEmitter->emit('\\\\OC\\\\User', 'postCreateUser', Array)\n#8 \/var\/www\/html\/encryption_vincent\/lib\/private\/User\/Manager.php(298): OC\\Hooks\\PublicEmitter->emit('\\\\OC\\\\User', 'postCreateUser', Array)\n#9 \/var\/www\/html\/encryption_vincent\/apps\/provisioning_api\/lib\/Users.php(163): OC\\User\\Manager->createUser('subadmin', '123456')\n#10 \/var\/www\/html\/encryption_vincent\/lib\/private\/legacy\/api.php(177): OCA\\Provisioning_API\\Users->addUser(Array)\n#11 [internal function]: OC_API::call(Array)\n#12 \/var\/www\/html\/encryption_vincent\/lib\/private\/Route\/Router.php(299): call_user_func(Array, Array)\n#13 \/var\/www\/html\/encryption_vincent\/ocs\/v1.php(61): OC\\Route\\Router->match('\/ocs\/cloud\/user...')\n#14 {main}","File":"\/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/KeyManager.php","Line":545}
[Thu Jan 12 08:18:44 2017] Failed addUser attempt with exception: Public Key missing for user: subadmin
```

PVince81 commented:
@SergioBertolinSG yeah I saw that, but it doesn't seem to affect the test outcome.

SergioBertolinSG commented:
My local failures:

/var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:269
/var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:453
/var/www/html/encryption_vincent/build/integration/features/webdav-related.feature:57
/var/www/html/encryption_vincent/build/integration/features/webdav-related.feature:111
/var/www/html/encryption_vincent/build/integration/sharees_features/sharees_provisioningapiv2.feature:9
 /var/www/html/encryption_vincent/build/integration/sharees_features/sharees_provisioningapiv2.feature:25

```
/var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:269

And Downloading file "/textfile0 (2).txt" with range "bytes=1-7"                 # FeatureContext::downloadFileWithRange()
      Server error response [url] http://localhost:8080/remote.php/webdav/textfile0%20(2).txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)
[Thu Jan 12 08:23:12 2017] 127.0.0.1:49722 [500]: /remote.php/webdav/textfile0%20(2).txt


/var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:453

When Downloading file "/common/sub/textfile0.txt" with range "bytes=9-17"                                   # FeatureContext::downloadFileWithRange()
      Server error response [url] http://localhost:8080/remote.php/webdav/common/sub/textfile0.txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)
[Thu Jan 12 08:25:34 2017] 127.0.0.1:50027 [500]: /remote.php/webdav/common/sub/textfile0.txt



    /var/www/html/encryption_vincent/build/integration/features/webdav-related.feature:57

And Downloaded content when downloading file "/testshare/overwritethis.txt" with range "bytes=0-6" should be "Welcome" # FeatureContext::downloadedContentWhenDownloadindShouldBe()
      Server error response [url] http://localhost:8080/remote.php/webdav/testshare/overwritethis.txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)[Thu Jan 12 08:41:04 2017] 127.0.0.1:52245 [500]: /remote.php/webdav/testshare/overwritethis.txt




    /var/www/html/encryption_vincent/build/integration/features/webdav-related.feature:111

    And Downloaded content when downloading file "/testshare/overwritethis.txt" with range "bytes=0-6" should be "Welcome" # FeatureContext::downloadedContentWhenDownloadindShouldBe()
      Server error response [url] http://localhost:8080/remote.php/webdav/testshare/overwritethis.txt [status code] 500 [reason phrase] Internal Server Error (GuzzleHttp\Exception\ServerException)
[Thu Jan 12 08:41:39 2017] 127.0.0.1:52314 [500]: /remote.php/webdav/testshare/overwritethis.txt



    /var/www/html/encryption_vincent/build/integration/sharees_features/sharees_provisioningapiv2.feature:9

And user "Sharee1" exists                           # ShareesContext::assureUserExists()
      Client error response [url] http://localhost:8080/ocs/v2.php/cloud/users [status code] 400 [reason phrase] Bad Request (GuzzleHttp\Exception\ClientException)



    /var/www/html/encryption_vincent/build/integration/sharees_features/sharees_provisioningapiv2.feature:25

    And user "Sharee1" exists                           # ShareesContext::assureUserExists()
      Client error response [url] http://localhost:8080/ocs/v2.php/cloud/users [status code] 400 [reason phrase] Bad Request (GuzzleHttp\Exception\ClientException)
```

Last two are related with https://github.com/owncloud/core/pull/26917

Test runned using alt home and encryption.

PVince81 commented:
Yes. Others are related to https://github.com/owncloud/core/pull/26844#issuecomment-271958351

SergioBertolinSG commented:
Yes, I guess user1 can rename it. Do you want that I add fixes here?

PVince81 commented:
@SergioBertolinSG yes, please add them here. Thanks

SergioBertolinSG commented:
When renaming I am getting an exception:

```
Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.
[Thu Jan 12 10:00:48 2017] Exception: {"Message":"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.","Exception":"OC\\Encryption\\Exceptions\\DecryptionFailedException","Code":0,"Trace":"#0 \/var\/www\/html\/encryption_vincent\/lib\/private\/Files\/Stream\/Encryption.php(459): OCA\\Encryption\\Crypto\\Encryption->decrypt('QFSVoU6JUkXVK4J...', '0end')\n#1 \/var\/www\/html\/encryption_vincent\/lib\/private\/Files\/Stream\/Encryption.php(290): OC\\Files\\Stream\\Encryption->readCache()\n#2 [internal function]: OC\\Files\\Stream\\Encryption->stream_read(23)\n#3 \/var\/www\/html\/encryption_vincent\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/Wrapper.php(83): fread(Resource id #57, 8192)\n#4 \/var\/www\/html\/encryption_vincent\/apps\/files_external\/3rdparty\/icewind\/streams\/src\/CallbackWrapper.php(91): Icewind\\Streams\\Wrapper->stream_read(8192)\n#5 [internal function]: Icewind\\Streams\\CallbackWrapper->stream_read(8192)\n#6 \/var\/www\/html\/encryption_vincent\/lib\/composer\/sabre\/http\/lib\/Sapi.php(78): stream_copy_to_stream(Resource id #60, Resource id #65, '7')\n#7 \/var\/www\/html\/encryption_vincent\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(498): Sabre\\HTTP\\Sapi::sendResponse(Object(Sabre\\HTTP\\Response))\n#8 \/var\/www\/html\/encryption_vincent\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#9 \/var\/www\/html\/encryption_vincent\/apps\/dav\/appinfo\/v1\/webdav.php(57): Sabre\\DAV\\Server->exec()\n#10 \/var\/www\/html\/encryption_vincent\/remote.php(164): require_once('\/var\/www\/html\/e...')\n#11 {main}","File":"\/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/Crypto\/Encryption.php","Line":360,"User":"user3"}

```

SergioBertolinSG committed:
Fix for Reshared files can be still accessed if a user in the middle removes it.

SergioBertolinSG commented:
After cherry picking commits from https://github.com/owncloud/core/pull/26917 (7327fa9 730d6ae 8b669b8 1e1c11f)

The renaming has worked, but the output is full of this exception:

```
Backends provided no user object for user0
[Thu Jan 12 12:16:16 2017] Exception: {"Exception":"OC\\User\\NoUserException","Message":"Backends provided no user object for user0","Code":0,"Trace":"#0 \/var\/www\/html\/encryption_vincent\/lib\/private\/Encryption\/Keys\/Storage.php(348): OC\\Files\\Filesystem::initMountPoints('user0')\n#1 \/var\/www\/html\/encryption_vincent\/lib\/private\/Encryption\/Keys\/Storage.php(181): OC\\Encryption\\Keys\\Storage->setupUserMounts('user0')\n#2 \/var\/www\/html\/encryption_vincent\/lib\/private\/Encryption\/Keys\/Storage.php(140): OC\\Encryption\\Keys\\Storage->constructUserKeyPath('OC_DEFAULT_MODU...', 'publicKey', 'user0')\n#3 \/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/KeyManager.php(586): OC\\Encryption\\Keys\\Storage->deleteUserKey('user0', 'publicKey', 'OC_DEFAULT_MODU...')\n#4 \/var\/www\/html\/encryption_vincent\/apps\/encryption\/lib\/Hooks\/UserHooks.php(200): OCA\\Encryption\\KeyManager->deletePublicKey('user0')\n#5 \/var\/www\/html\/encryption_vincent\/lib\/private\/legacy\/hook.php(105): OCA\\Encryption\\Hooks\\UserHooks->postDeleteUser(Array)\n#6 \/var\/www\/html\/encryption_vincent\/lib\/private\/Server.php(261): OC_Hook::emit('OC_User', 'post_deleteUser', Array)\n#7 [internal function]: OC\\Server->OC\\{closure}(Object(OC\\User\\User))\n#8 \/var\/www\/html\/encryption_vincent\/lib\/private\/Hooks\/EmitterTrait.php(98): call_user_func_array(Object(Closure), Array)\n#9 \/var\/www\/html\/encryption_vincent\/lib\/private\/Hooks\/PublicEmitter.php(32): OC\\Hooks\\BasicEmitter->emit('\\\\OC\\\\User', 'postDelete', Array)\n#10 \/var\/www\/html\/encryption_vincent\/lib\/private\/User\/User.php(229): OC\\Hooks\\PublicEmitter->emit('\\\\OC\\\\User', 'postDelete', Array)\n#11 \/var\/www\/html\/encryption_vincent\/apps\/provisioning_api\/lib\/Users.php(348): OC\\User\\User->delete()\n#12 \/var\/www\/html\/encryption_vincent\/lib\/private\/legacy\/api.php(177): OCA\\Provisioning_API\\Users->deleteUser(Array)\n#13 [internal function]: OC_API::call(Array)\n#14 \/var\/www\/html\/encryption_vincent\/lib\/private\/Route\/Router.php(299): call_user_func(Array, Array)\n#15 \/var\/www\/html\/encryption_vincent\/ocs\/v1.php(61): OC\\Route\\Router->match('\/ocs\/cloud\/user...')\n#16 {main}","File":"\/var\/www\/html\/encryption_vincent\/lib\/private\/Files\/Filesystem.php","Line":413}
```

SergioBertolinSG committed:
Fixed scenario getting all shares of a user using that user

SergioBertolinSG committed:
Fixed scenario getting share info of a share

SergioBertolinSG commented:
Running tests in current status with alt home and encryption there are still two failures:

    /var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:219
    /var/www/html/encryption_vincent/build/integration/features/sharing-v1.feature:284

SergioBertolinSG commented:
Ok now those tests pass. @PVince81 I'll run the full suite and put result here. 
Before having this finished we need to address that exception https://github.com/owncloud/core/pull/26844#issuecomment-272153302 .

SergioBertolinSG commented:
Tests has passed locally.

PVince81 commented:
Rebased

PVince81 commented:
@SergioBertolinSG please try with this fix https://github.com/owncloud/core/pull/26938, the errors should disappear.

SergioBertolinSG commented:
It works with fine 4559dad3e421f330eb8cbc718aa19e573a5ae2f7

PVince81 commented:
https://github.com/owncloud/core/pull/26938 was merged. I have rebased to include it.

@SergioBertolinSG rerun all tests locally ? Hopefully they'll all pass now.

SergioBertolinSG commented:
Test run locally, they worked fine. (Both with encryption and alt home and without).

PVince81 commented:
Rebased for CI

PVince81 commented:
WTF, now unrelated unit tests fail

> Test\Encryption\Keys\StorageTest::testGetUserKeyWhenKeyStorageIsOutsideHome
mounting was not necessary
Failed asserting that true is false.

> /var/lib/jenkins/workspace/owncloud-core_core_PR-26844-75FHI25UJEAZQY24FCAWRPW7SHNMBAWRD4CUOZ4EEGK72KF2R3HA/tests/lib/Encryption/Keys/StorageTest.php:310

PVince81 commented:
Reproduceable locally, I'll have a look

PVince81 commented:
Removed old bogus encryption fix that we cherry-picked onto this branch for testing but that is obsolete and fixed differently on master.

👍 for @SergioBertolinSG's changes.

Let's get this in.

PVince81 commented:
stable9.1: https://github.com/owncloud/core/pull/26993
stable9: TODO

SergioBertolinSG commented:
Requires to backport first https://github.com/owncloud/core/pull/26846, https://github.com/owncloud/core/pull/26920 and https://github.com/owncloud/core/pull/26865
